### ISSUE TITLE
model doesn't learn/weights don't get updated

### ISSUE THREAD

**Issue by daikazoku** [fchollet/keras#255]
I need to start by saying that I'm just toying with machine learning, so this could just be me doing stupid mistakes.

I wrote a very small and useless model just to check that my script works, since I'm on a cpu. I keep giving it the same batch just to see some quick change in weights, but they don't change anyway.
This seem to happen with any architecture. I started with a smaller version of the VGG example.
I've tried with every optimizer, this version of the script uses RMSprop because it shouldn't matter if I'm choosing bad hyperparameters.

The model has the exact same loss at every step of training.

```
2.1586735246819186 0.9375
32/32 [==============================] - 1s - loss: 11.8727 - acc.: 0.6562
2.1586735246819186 0.9375
32/32 [==============================] - 1s - loss: 11.8727 - acc.: 0.6562
...
2.1586735246819186 0.9375
32/32 [==============================] - 1s - loss: 11.8727 - acc.: 0.6562
2.1586735246819186 0.9375
32/32 [==============================] - 1s - loss: 11.8727 - acc.: 0.6562
```

I save at each "epoch" (in this case single batch) the filters.

This is the first filter before the first epoch, just initialized.
![filter1](https://cloud.githubusercontent.com/assets/11365004/8270827/fe01ff24-17f6-11e5-8b06-1512bdb7310b.png)

First filter after ~10 epochs.
![filter37](https://cloud.githubusercontent.com/assets/11365004/8270828/12683b5e-17f7-11e5-885b-76b6659546fe.png)

This is the script I'm using:
```
fsize = 15

model = Sequential()
model.add(Convolution2D(4, 3, fsize, fsize)) 
model.add(Activation('relu'))
model.add(MaxPooling2D(poolsize=(imsize - fsize + 1, imsize - fsize + 1)))
model.add(Flatten())

model.add(Dense(4, 2))
model.add(Activation('softmax'))

optimizer = RMSprop()
model.compile(loss='categorical_crossentropy', optimizer=optimizer)


# ! save filters
w = model.get_weights()[0]
j = 0
for f in w:
	j += 1
	f = np.swapaxes(f, 0, 2)
	f = np.swapaxes(f, 0, 1)
	misc.imsave("filter" + str(j) + ".png", f)
# / save filters

for epoch in range(200):
	for batch in range(1):
		#loadTrain(a, b) loads b examples starting from a
		X_train, Y_train = loadTrain(32 * batch + 200, 32)

		# this has the same result:
		# model.fit(X_train, Y_train, nb_epoch=1)

		loss, accuracy = model.train(X_train, Y_train, accuracy=True)
		print(loss, accuracy)
		

	X_test, Y_test = loadTrain(1100, 32)
	model.evaluate(X_test, Y_test, show_accuracy=True)

	# ! save filters
	w = model.get_weights()[0]
	for f in w:
		j += 1
		f = np.swapaxes(f, 0, 2)
		f = np.swapaxes(f, 0, 1)
		misc.imsave("filter" + str(j) + ".png", f)

	# / save weights
```

**Comment by mthrok** [comment]
The way you apply `MaxPooling` looks somewhat peculiar to me.
How about this?
```
model.add(Convolution2D(4, 3, fsize, fsize, border_mode='full')) 
model.add(Activation('relu'))
model.add(Convolution2D(4, 4, fsize, fsize))
model.add(Activation('relu'))
model.add(MaxPooling2D(poolsize=(2, 2)))

model.add(Flatten())
model.add(Dense(4*(imsize/2)*(imsize/2), 2))
```

**Comment by daikazoku** [comment]
This seems to learn something for an epoch, but then stops learning again. The filters change from uninitialized to first epoch, but then stay identical.
```
16.849789491436628 0.0625
32/32 [==============================] - 5s - loss: 11.8727 - acc.: 0.6562
2.1586735246819186 0.9375
32/32 [==============================] - 5s - loss: 11.8727 - acc.: 0.6562
2.1586735246819186 0.9375
32/32 [==============================] - 5s - loss: 11.8727 - acc.: 0.6562
2.1586735246819186 0.9375
...
```

```
$ md5sum filter1.png #filter 1, epoch 1
5dd44cb092194e5792c33e77d4dfe431  filter1.png
$ md5sum filter5.png #filter 1, epoch 2
2844392fc88c7a0b270eb2fd56087646  filter5.png
$ md5sum filter9.png #filter 1, epoch 3
2844392fc88c7a0b270eb2fd56087646  filter9.png
```

**Comment by mthrok** [comment]
```python
for epoch in range(200):
    for batch in range(1):
        #loadTrain(a, b) loads b examples starting from a
        X_train, Y_train = loadTrain(32 * batch + 200, 32)
```
If you use different batch on each epoch, what happens?

**Comment by mthrok** [comment]
One more thing, are the labels in `Y_train` shuffled?

**Comment by daikazoku** [comment]
If I use more batches, the same thing happens:

```
1.95735197403513 0.6875
28.062755820864915 0.1875
9.99200722162641e-16 1.0
11.872704385750545 0.65625
19.42806172213725 0.4375
32/32 [==============================] - 5s - loss: 11.8727 - acc.: 0.6562
2.1586735246819186 0.9375
28.062755820864915 0.1875
9.99200722162641e-16 1.0
11.872704385750545 0.65625
19.42806172213725 0.4375
32/32 [==============================] - 5s - loss: 11.8727 - acc.: 0.6562
...
```

Except for the first step batch in the first and second epoch, the loss seems to always be identical.

I'm not sure I understand what you mean by shuffled. There are some runs of (0, 1) examples and (1, 0) examples. The loss is identical to the last digit, so I don't think that's the actual problem. Do I try shuffling the order of the images in the dataset anyway? (if I just get images at random from the dataset, we won't be able to see the identical loss problem)

**Comment by daikazoku** [comment]
except the first batch of the first epoch, the loss is identical across runs:

```
20.678397291060108 0.0625
28.062755820864915 0.1875
9.99200722162641e-16 1.0
11.872704385750545 0.65625
19.42806172213725 0.4375
32/32 [==============================] - 5s - loss: 11.8727 - acc.: 0.6562
2.1586735246819186 0.9375
28.062755820864915 0.1875
9.99200722162641e-16 1.0
```

**Comment by mthrok** [comment]
By 'shuffled',  I meant if the label is not ordered by class or not. 
In binary classification case, `Y_train` must look like
```
[[ 0, 1],
 [ 1, 0],
 [ 0, 1]]
```

**Comment by mthrok** [comment]
Did you try `cifar10_cnn.py` example and did it produce meaningful result?

**Comment by daikazoku** [comment]
The cifar10 example's loss is slowly but steadily decreasing, so it works.

Unfortunately, still same problem after shuffling the label so that they are always like in your example:

```
0.2355973707751222 0.96875
17.269388197455335 0.5
17.269388197455335 0.5
17.269388197455335 0.5
17.269388197455335 0.5
32/32 [==============================] - 5s - loss: 17.2694 - acc.: 0.5000
17.269388197455335 0.5
17.269388197455335 0.5
...
```

**Comment by mthrok** [comment]
What is your dataset like?

**Comment by daikazoku** [comment]
It's a toy dataset from a 3d computer game. In the first class, there is some player other than the local player in the image. In the second class, there is only the local player in the image. There are about 1000 images, 500 per class. They are screenshots taken sequentially and then manually classified. They are loaded with labels
```
[[ 1.  0.]
 [ 0.  1.]
 [ 1.  0.]
 [ 0.  1.]
...
 [ 1.  0.]
 [ 0.  1.]
 [ 1.  0.]
 [ 0.  1.]]
```
as you suggested.

**Comment by daikazoku** [comment]
The problem was my dataset. When normalizing each pixel to [0, 1], the array was still int, and each pixel bcame 0.

Now that I properly normalize the images, the model is learning.

Sorry to have wasted your time.

**Comment by mthrok** [comment]
Glad you solved it.
